File: r152/src/extree152.py
#!/usr/bin/env python3
"""Round 152 -- the VALIDATOR for explicit exhaustion trees (L6-EXTREE-1).

This program performs NO SEARCH.  It walks the tree written in the file and, at
every node, does three things:

  1. recomputes the legal moves from the state itself and insists the file's
     N<k> lists EXACTLY that many children, then descends into them in the
     catalogue's canonical order -- so a subtree cannot be quietly dropped;
  2. for every leaf, checks that at least one of three PROVED facts justifies
     stopping there:
        (b) ports + (120 - |hexagons used|) + (a + bb + e left) < C+1   [(P2)]
        (f) the feasibility lemma already excludes every continuation   [r150]
        (p) ports + ub(remaining budgets) - 1 < C+1, where ub is the monotone
            bound from cells validated EARLIER IN THIS SAME FILE         [(P1)]
  3. checks the node does not itself reach C+1 ports at deficit <= d.

If all three hold everywhere, the tree is a complete case analysis and
cap(K) <= C.  The order of the file is the induction order: a cell may only be
used in a (p) justification after its own tree has been validated.

The catalogue comes from r152/src/checker152.py, which builds it from string
algebra and reads nothing from the production solvers.

usage: extree152.py --tree <file.extree> [--report r.json] [--cert cap_cert.txt]
"""
from __future__ import annotations
import argparse, hashlib, json, sys, time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from checker152 import (HEX, ORB, PHASE, MOVES, NHEX, Checker,          # noqa
                        read_text_cert, entries_of)

KIND_ORDER = {"E": 0, "A": 1, "B": 2, "P": 3, "H": 4}


def canonical_moves(v):
    """The catalogue's canonical order: clean E, dirty A, dirty B, paid, heavy."""
    return sorted(MOVES[v], key=lambda m: (KIND_ORDER[m[1]],))


_CANON = None


def canon(v):
    global _CANON
    if _CANON is None:
        _CANON = [canonical_moves(u) for u in range(len(MOVES))]
    return _CANON[v]


class Validator:
    def __init__(self, certified):
        self.cert = certified
        self._ubc = {}
        self.nodes = 0

    def ub(self, tok, d, a, bb, e, h):
        key = (tok, d, a, bb, e, h)
        v = self._ubc.get(key)
        if v is not None:
            return v
        best = 120 + a + bb + e
        for (kb, kd, ka, kbb, ke, kh), c in self.cert.items():
            if (kb >= tok and kd >= d and ka >= a and kbb >= bb
                    and ke >= e and kh >= h and c < best):
                best = c
        self._ubc[key] = best
        return best

    @staticmethod
    def feas(phm, corb, tok, dmax):
        us = sorted((5 - len(ph) for q, ph in phm.items() if q != corb),
                    reverse=True)
        return sum(us[tok:]) <= dmax

    def validate(self, cell, cap, toks, pos):
        """Walk one tree.  Returns (ok, detail, next position)."""
        b, dmax, amax, bmax, emax, hmax = cell
        phm = {ORB[0]: {PHASE[0]}}
        hexc = {HEX[0]: 1}
        state = dict(err=None, pos=pos)

        def legal(v, corb, tok, au, bu, eu, hu):
            out = []
            for t, kind, cost in canon(v):
                q = ORB[t]
                ph = phm.get(q)
                if ph is not None and PHASE[t] in ph:
                    continue
                if kind == "E" and q != corb:
                    continue
                if kind == "A" and au >= amax:
                    continue
                if kind == "B" and bu >= bmax:
                    continue
                if kind == "H" and hu + cost > hmax:
                    continue
                newhex = HEX[t] not in hexc
                se = 0
                if not newhex and kind not in ("A", "B"):
                    if eu >= emax:
                        continue
                    se = 1
                fresh = ph is None
                c = 0 if (kind == "E" or fresh) else 1
                if c > tok:
                    continue
                out.append((t, kind, cost, fresh, newhex, se, c))
            return out

        def walk(v, corb, ports, tok, au, bu, eu, hu, deficit):
            self.nodes += 1
            i = state["pos"]
            if i >= len(toks):
                state["err"] = "token stream ended early"
                return
            tk = toks[i]
            state["pos"] = i + 1
            if ports >= cap + 1 and deficit <= dmax:
                state["err"] = (f"the tree itself reaches {ports} ports at "
                                f"deficit {deficit} <= {dmax}")
                return
            if tk == "L":
                r1 = ports + (NHEX - len(hexc)) + (amax - au) + (bmax - bu) \
                     + (emax - eu)
                if r1 < cap + 1:
                    return                                            # (b)
                if not self.feas(phm, corb, tok, dmax):
                    return                                            # (f)
                r2 = ports + self.ub(tok, dmax - deficit + 4 + 5 * tok,
                                     amax - au, bmax - bu, emax - eu,
                                     hmax - hu) - 1
                if r2 < cap + 1:
                    return                                            # (p)
                state["err"] = (f"unjustified leaf at token {i}: ports={ports} "
                                f"deficit={deficit} r1={r1} r2={r2}")
                return
            if not tk.startswith("N"):
                state["err"] = f"bad token {tk!r} at {i}"
                return
            k = int(tk[1:])
            ms = legal(v, corb, tok, au, bu, eu, hu)
            if k != len(ms):
                state["err"] = (f"token {i} says {k} children, the state has "
                                f"{len(ms)} legal moves")
                return
            for t, kind, cost, fresh, newhex, se, c in ms:
                q = ORB[t]
                if fresh:
                    phm[q] = {PHASE[t]}
                else:
                    phm[q].add(PHASE[t])
                hexc[HEX[t]] = hexc.get(HEX[t], 0) + 1
                walk(t, q, ports + 1, tok - c, au + (kind == "A"),
                     bu + (kind == "B"), eu + se,
                     hu + (cost if kind == "H" else 0),
                     deficit + (4 if fresh else -1))
                hexc[HEX[t]] -= 1
                if hexc[HEX[t]] == 0:
                    del hexc[HEX[t]]
                if fresh:
                    del phm[q]
                else:
                    phm[q].discard(PHASE[t])
                if state["err"]:
                    return

        sys.setrecursionlimit(10000)
        walk(0, ORB[0], 1, b, 0, 0, 0, 0, 4)
        return state["err"] is None, state["err"], state["pos"]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--tree", required=True)
    ap.add_argument("--cert", default=None,
                    help="cap certificate, to also replay the witnesses")
    ap.add_argument("--report", default="r152/certs/extree_152.json")
    a = ap.parse_args()
    raw = Path(a.tree).read_bytes()
    text = raw.decode()
    if not text.startswith("L6-EXTREE-1"):
        raise SystemExit("not an L6-EXTREE-1 file")
    toks = [t for line in text.splitlines()
            for t in line.split("#")[0].split()
            if not t.startswith("L6-EXTREE")]

    wit = {}
    if a.cert:
        doc = read_text_cert(Path(a.cert).read_text())
        for e in entries_of(doc):
            wit["|".join(map(str, e["args"])) + f"@{e['cap']}"] = e["witness"]

    certified, rows, ok, i = {}, [], True, 0
    total_nodes = 0
    while i < len(toks):
        if toks[i] != "tree":
            raise SystemExit(f"expected 'tree' at token {i}, saw {toks[i]!r}")
        cell = tuple(int(x) for x in toks[i + 1:i + 7])
        cap = int(toks[i + 7])
        key = "|".join(map(str, cell))
        v = Validator(dict(certified))
        t0 = time.time()
        good, err, i = v.validate(cell, cap, toks, i + 8)
        row = dict(cell=key, cap=cap, nodes=v.nodes,
                   seconds=round(time.time() - t0, 2))
        if not good:
            row.update(status="TREE_INVALID", detail=err)
        else:
            row["status"] = "UPPER_CERTIFIED"
            certified[cell] = cap
            w = wit.get(f"{key}@{cap}")
            if w:
                g, info = Checker({}, 0).replay(cell, w)
                if g and info["ports"] == cap:
                    row["status"] = "EXACT_CERTIFIED"
                else:
                    row.update(status="WITNESS_BAD", detail=info)
        if row["status"] not in ("EXACT_CERTIFIED", "UPPER_CERTIFIED"):
            ok = False
        total_nodes += v.nodes
        rows.append(row)
        print(f"  {key:>16} cap={cap:<4} {row['status']:<16} "
              f"nodes={v.nodes:<12,} {row['seconds']}s"
              + (f"  {row.get('detail')}" if row.get("detail") else ""),
              flush=True)
    out = dict(tree=a.tree, tree_sha256=hashlib.sha256(raw).hexdigest(),
               tree_bytes=len(raw), cells=len(rows), total_nodes=total_nodes,
               certified=sum(r["status"] in ("EXACT_CERTIFIED",
                                             "UPPER_CERTIFIED") for r in rows),
               exact=sum(r["status"] == "EXACT_CERTIFIED" for r in rows),
               all_valid=ok, rows=rows)
    Path(a.report).parent.mkdir(parents=True, exist_ok=True)
    Path(a.report).write_text(json.dumps(out, indent=1) + "\n")
    print(json.dumps({k: v for k, v in out.items() if k != "rows"}, indent=1))
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: r152/src/extree152.c
/* Round 152 -- producer of EXPLICIT EXHAUSTION TREES (format L6-EXTREE-1).
 *
 * The upper-bound direction cap(K) <= C is a complete case analysis, and this
 * writes that case analysis down instead of asking a reader to rerun a search.
 * The file is a preorder token stream, one token per node of the tree:
 *
 *     N<k>   an internal node whose children are EXACTLY the k legal moves
 *            from its state, in the catalogue's canonical order
 *     L      a leaf, justified by one of three PROVED facts (the checker works
 *            out which, it is not told):
 *              (b) ports + (120 - |hexagons used|) + (a+bb+e left) < C+1
 *              (f) the feasibility lemma already excludes the continuation
 *              (p) the (P1) monotone bound from a cell certified EARLIER in
 *                  this same file
 *
 * A reader verifies it with r152/src/extree152.py, which performs NO search:
 * it recomputes the legal moves at each node, insists the file lists all of
 * them, and checks each leaf's justification.
 *
 * build: cc -O2 -o extree152.exe extree152.c
 * run:   ./extree152.exe <cap_cert.txt> <out.extree> [node_cap]
 */
#define CHECKER152_NO_MAIN
#include "checker152.c"

static FILE *OUT;
static uint64_t TNODES;
static int ECAP;                     /* the claimed cap of the current cell */

/* the same legality filter the checker will apply, in the same order */
static int legal_moves(int cur, int corb, int tok, int au, int bu, int eu,
                       int hu, int *tgt, int *knd, int *cst) {
    int n = 0;
    static int cand[NP + 16], kind[NP + 16], cost[NP + 16];
    cand[0] = FREE_T[cur]; kind[0] = 4; cost[0] = 0;
    cand[1] = DA_T[cur];   kind[1] = 1; cost[1] = 0;
    cand[2] = DB_T[cur];   kind[2] = 2; cost[2] = 0;
    int m = 3;
    for (int i = 0; i < 5; ++i) { cand[m] = PAID_T[cur][i]; kind[m] = 0; cost[m] = 0; ++m; }
    for (int i = 0; i < NHV[cur]; ++i) {
        cand[m] = HV_T[cur][i]; kind[m] = 3; cost[m] = HV_C[cur][i]; ++m;
        if (m >= NP + 16) { fprintf(stderr, "move overflow\n"); exit(3); }
    }
    for (int i = 0; i < m; ++i) {
        int t = cand[i], k = kind[i], c = cost[i];
        if (k == 1 && au >= CA) continue;
        if (k == 2 && bu >= CBB) continue;
        if (k == 3 && hu + c > CH) continue;
        int q = ORBID[t];
        if (phm[q] >> PHASE[t] & 1) continue;
        if (k == 4 && q != corb) continue;
        int newhex = !hexu[HEXID[t]];
        if (!newhex && k != 1 && k != 2 && eu >= CE) continue;
        int fresh = (phm[q] == 0);
        int tc = (k == 4 || fresh) ? 0 : 1;
        if (tc > tok) continue;
        tgt[n] = t; knd[n] = k; cst[n] = c; ++n;
    }
    return n;
}

static int leaf_reason(int corb, int ports, int tok, int au, int bu, int eu, int hu) {
    if (ports + (NH - hexcount) + (CA - au) + (CBB - bu) + (CE - eu) < ECAP + 1)
        return 'b';
    if (!feas(tok, corb)) return 'f';
    if (ports + ubound(tok, CD - deficit + 4 + 5 * tok, CA - au, CBB - bu,
                       CE - eu, CH - hu) - 1 < ECAP + 1)
        return 'p';
    return 0;
}

static void emit(int cur, int corb, int ports, int tok, int au, int bu, int eu, int hu) {
    ++TNODES;
    if (NODECAP && TNODES > NODECAP) { CAPPED = 1; return; }
    if (ports >= ECAP + 1 && deficit <= CD) { FOUND = 1; return; }
    if (leaf_reason(corb, ports, tok, au, bu, eu, hu)) { fputs("L ", OUT); return; }
    int *tgt = malloc(sizeof(int) * 3 * (NP + 16));
    int *knd = tgt + (NP + 16), *cst = knd + (NP + 16);
    if (!tgt) { fprintf(stderr, "alloc\n"); exit(3); }
    int n = legal_moves(cur, corb, tok, au, bu, eu, hu, tgt, knd, cst);
    fprintf(OUT, "N%d ", n);
    for (int i = 0; i < n && !FOUND && !CAPPED; ++i) {
        int t = tgt[i], k = knd[i], c = cst[i], q = ORBID[t];
        int fresh = (phm[q] == 0), newhex = !hexu[HEXID[t]];
        int spend_e = (!newhex && k != 1 && k != 2) ? 1 : 0;
        int tc = (k == 4 || fresh) ? 0 : 1;
        unsigned char old = phm[q];
        phm[q] = (unsigned char)(old | (1u << PHASE[t]));
        if (fresh) { opened[nopened++] = q; deficit += 4; } else deficit -= 1;
        if (newhex) { hexu[HEXID[t]] = 1; ++hexcount; }
        emit(t, q, ports + 1, tok - tc, au + (k == 1), bu + (k == 2),
             eu + spend_e, hu + (k == 3 ? c : 0));
        if (newhex) { hexu[HEXID[t]] = 0; --hexcount; }
        if (fresh) { --nopened; deficit -= 4; } else deficit += 1;
        phm[q] = old;
    }
    free(tgt);
}

int main(int argc, char **argv) {
    if (argc < 3) { fprintf(stderr, "usage: %s <cap_cert.txt> <out.extree> [node_cap]\n", argv[0]); return 3; }
    NODECAP = argc > 3 ? strtoull(argv[3], NULL, 10) : 0;
    common_setup();
    read_cert(argv[1]);
    OUT = fopen(argv[2], "w");
    if (!OUT) { fprintf(stderr, "cannot write %s\n", argv[2]); return 3; }
    fputs("L6-EXTREE-1\n", OUT);
    fputs("# preorder: N<k> = internal with exactly k legal children,\n", OUT);
    fputs("# L = leaf justified by (b), (f) or (p); see extree152.py\n", OUT);
    const int bad = 0;
    for (int i = 0; i < NCELL; ++i) {
        Cell *c = &CELLS[i];
        CB = c->arg[0]; CD = c->arg[1]; CA = c->arg[2];
        CBB = c->arg[3]; CE = c->arg[4]; CH = c->arg[5];
        ECAP = c->cap;
        TNODES = 0; CAPPED = 0; FOUND = 0;
        memset(phm, 0, sizeof phm); memset(hexu, 0, sizeof hexu);
        nopened = 0; hexcount = 1; hexu[HEXID[0]] = 1;
        phm[ORBID[0]] = (unsigned char)(1u << PHASE[0]);
        opened[nopened++] = ORBID[0]; deficit = 4;
        fprintf(OUT, "\ntree %d %d %d %d %d %d %d\n", c->arg[0], c->arg[1],
                c->arg[2], c->arg[3], c->arg[4], c->arg[5], c->cap);
        emit(0, ORBID[0], 1, CB, 0, 0, 0, 0);
        fputc('\n', OUT);
        if (FOUND || CAPPED) {
            /* The stream for this cell is truncated, so the whole file is
             * unusable.  Say so and stop rather than leave a file that looks
             * complete. */
            fprintf(stderr, "cell %d|%d|%d|%d|%d|%d: %s -- aborting, %s is "
                    "incomplete\n", c->arg[0], c->arg[1], c->arg[2], c->arg[3],
                    c->arg[4], c->arg[5], FOUND ? "DISAGREE" : "node cap",
                    argv[2]);
            fclose(OUT);
            return 1;
        }
        ub_add(c->arg, c->cap);
        fprintf(stderr, "%4d/%4d %d|%d|%d|%d|%d|%d cap=%d nodes=%llu\n", i + 1,
                NCELL, c->arg[0], c->arg[1], c->arg[2], c->arg[3], c->arg[4],
                c->arg[5], c->cap, (unsigned long long)TNODES);
    }
    fclose(OUT);
    fprintf(stderr, "cells=%d trees_written=%d skipped=%d\n", NCELL, NCELL - bad, bad);
    return bad ? 1 : 0;
}
